More robust timing tests

gilles-peskine-arm opened the pull request:
The primary objective of this PR is to fix the issue we have with `test_suite_timing` and the TIMING selftest on the CI, which sometimes fail due to heavy load. I also fixed some minor bugs in the timing module.

This PR introduces new timing tests that test the functions in the timing module mostly independently. The old self-test is still present in the code but will not be executed by the test suite, like other self-tests that we judge to be too time-consuming. CI still runs it once through `programs/test/selftest`.

Compared with the self-tests:

* The new tests are more modular and print some detailed information when they fail.
* I don't test `mbedtls_set_alarm` with delays of more than 1 second, which makes the test run for about 2s instead of 8s.
* The timing self-test is more tolerant and shorter.

There are some test ideas I discarded in https://github.com/gilles-peskine-arm/mbedtls/tree/IOTSSL-1798-development-1 . The present PR is https://github.com/gilles-peskine-arm/mbedtls/tree/IOTSSL-1798-development-2 with a few minor fixes and a cleaned up history.

Incidental changes in this PR:

* Fix read from an uninitialized integer object in `mbedtls_timing_get_timer` on Windows (internal ref: IOTSSL-1822)
* Improved documentation of the timing API.
* Fix `mbedtls_set_alarm(0)` on Unix/POSIX.
* In the `selftest` program, support running a subset of the tests with command line arguments.

I've tested several iterations of this PR on my machine and on CI. The CI failures haven't completely gone so far, but they're markedly less frequent.

Partial backport for 2.1: #1223 (bug fixes, self test robustness, selftest command line parsing)  
Partial backport for 1.3: #1224 (bug fixes, self test robustness)

Internal ref: IOTSSL-1798

gilles-peskine-arm commented:
CI has now passed twice (I rebased a couple of times).

mazimkhan commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,359 @@
+    {
+        millis1 = mbedtls_timing_get_timer( &timer1, 0 );
+    }
+    while( millis1 < TIMING_SHORT_TEST_MS );
```
Same as above comment on missing ```iterations <= TIMING_SHORT_TEST_ITERATIONS_MAX```.

mazimkhan commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,359 @@
+           a busy environment such as an overloaded continuous integration
+           system, so we don't test it it. */
+    }
+    while( millis2 < TIMING_SHORT_TEST_MS );
```
Can we have a mechanism to break out of the loop in case of a bug. See previous comments.

mazimkhan commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,359 @@
+
+    /* Set an alarm and count how long it takes with a timer. */
+    (void) mbedtls_timing_get_timer( &timer, 1 );
+    mbedtls_set_alarm( seconds );
```
Should we not validate that ```seconds > 0```?

mazimkhan commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,359 @@
+        ++iterations;
+        millis = mbedtls_timing_get_timer( &timer, 0 );
+    }
+    while( !mbedtls_timing_alarmed && millis <= millis_max );
```
Here we are relying on ```mbedtls_timing_get_timer``` to work correctly. Watchdog can be based on ```iterations``` or in addition to the timer.

mazimkhan commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,359 @@
+    TEST_ASSERT( millis <= millis_max );
+
+    mbedtls_timing_alarmed = 0;
+    return;
```
Why is there a need for ```return```?

gilles-peskine-arm replied:
Because I don't want to print diagnosis information if the test succeeded.

mazimkhan commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,359 @@
+    int saw_status_1 = 0;
+    int warn_inconclusive = 0;
+
+    assert( int_ms >= 0 );
```
Why no test assert?

gilles-peskine-arm replied:
The test function expects `int_ms` and `fin_ms` to be `unsigned long`, but this type is not supported by the test framework, so I use `int` as the static type and put an additional dynamic type check to verify that the values are nonnegative. I used `TEST_ASSERT` for test failures, and `assert` for bad test input (like when there is an error parsing the `.data` file). Is this not correct?

mazimkhan replied:
From execution perspective ```TEST_ASSERT``` only aborts a test. ```assert``` will abort the whole suite as a suite built as one executable. Do you want to do that if one test has invalid data? Also this may not work as a wider strategy because data can be wrong mistakenly and a user may not put such checks.

mpg replied:
I don't have a strong opinion here, since either way tests with bad input will fail reliably and a few seconds of investigation will reveal that the problem is with the input data, and this should be a rare occurrence anyway.

I have however a slight preference for `TEST_ASSERT()` as it is just as clear (it's fairly obvious to readers that such a statement at the very top of the function is about input validity) and doesn't introduce anything new. But I can totally live with `assert()` too.

mazimkhan commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,359 @@
+       allows 250ms. */
+    delta1 = hardclock1 - hardclock0;
+    TEST_ASSERT( delta1 > 0 );
+    TEST_ASSERT( delta1 < 1000000000 );
```
Can we document why we have a big number ```1000000000 ``` although the delay was only 2 ms.

gilles-peskine-arm replied:
The previous comment is supposed to document this (“(…) If the hardclock counter runs at 4GHz, then one billion (which is 1/4 of the counter wraparound on a 32-bit machine) allows 250ms.”). Can you suggest how to clarify it?

gilles-peskine-arm commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,359 @@
+   first and run other test functions only if that timing_timer_simple
+   succeeded. */
+
+#include <limits.h>
```
`limits.h` is usually shipped with the compiler rather than with libc, so it's available everywhere (the C standard term is that it must be present on freestanding implementations). On Linux it comes with gcc.

mpg replied:
I addition to what Gilles said, we also use this header in the library (including in `check_config.h`) so if their was an issue with it, we would know about it.

gilles-peskine-arm commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,359 @@
+static int expected_delay_status( uint32_t int_ms, uint32_t fin_ms,
+                                  unsigned long actual_ms )
+{
+    return( fin_ms == 0 ? -1 :
```
Yes, `int_ms` is valid. We didn't document the API very well, but `int_ms=0` is a timer that expires immediately, whereas `fin_ms=0` is a diabled timer.

gilles-peskine-arm commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,359 @@
+    unsigned long millis = 0;
+    unsigned long iterations = 0;
+    /* Start the timer. */
+    (void) mbedtls_timing_get_timer( &timer, 1 );
```
Indeed I wouldn't have called a function to set a timer `get_timer`, but I'm only testing the API, not changing it.

gilles-peskine-arm commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,359 @@
+/* BEGIN_CASE */
+void timing_timer_reset( )
 {
-    TEST_ASSERT( mbedtls_timing_self_test( 1 ) == 0 );
```
These unit tests are completely independent from `mbedtls_timing_self_test`. We don't run all the self-tests in the unit tests, some of the more time-consuming ones are only run via `programs/test/selftest`. The timing test is the most time-consuming by far on a PC-class processor (it takes about 8s to run regardless of the platform — a fast PC can zip through all the other unit tests in less than that).

gilles-peskine-arm commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,359 @@
+        ++iterations;
+        millis = mbedtls_timing_get_timer( &timer, 0 );
+    }
+    while( millis < TIMING_SHORT_TEST_MS );
```
I tried to make a compromise between detecting infinite loops, and allowing very fast platforms where it takes a lot of iterations to reach the timeout. If `mbedtls_timing_get_timer` fails to increment `millis`, this should have been detected in the previous test case which calls `timing_timer_simple`.

I'm open to being convinced to do that in every test, but every such check is fragile (I need to hard-code a magic value for the iteration cap) and makes the test code more complex. So I prefer to rely on a higher-level timeout. I wasn't even sure about using a max iteration count in `timing_timer_simple`.

mazimkhan replied:
The check could be in addition to the time check hence not taking long if time is incremented correctly.
Although I understand the difficulty of getting a correct max count that can work with range of targets slow to fast. It would have been good if there was target specific multiplier to find no. of cycles. Even then it is not very portable due to varying instructions. Hence I will leave it up to you.

mpg replied:
Would it be possible to ensure that if `timing_timer_simple` fails, then all following tests fail before even running? I feel this would be a nice way to complete Gille's strategy of checking for infinite loops in one place only.

mazimkhan replied:
Execution of the tests is part of the test framework and it has look after execution requirements for all test suites. In this case it makes sense to abort the suite. But we may not want that in other suites. However, we can add a ```TEST_SUIT_ASSERT``` that aborts the whole suite.

gilles-peskine-arm commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,359 @@
+    while( millis < TIMING_SHORT_TEST_MS );
+
+    /* Reset the timer and check that it has restarted. */
+    TEST_ASSERT( mbedtls_timing_get_timer( &timer, 1 ) == 0 );
```
Well-spotted! Actually, `mbedtls_timing_get_timer` returns 0 on reset on Unix, but the old offset on Windows. I made the test consistent with Unix (that's why it passed on my machine), but the documentation consistent with Windows. I'll fix this in a follow-up commit.

gilles-peskine-arm commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,359 @@
+        TEST_ASSERT( saw_status_1 );
+    }
+
+    return;
```
If the test succeeds, I don't want to print the diagnosis information. The value of `warn_inconclusive` is used when printing the diagnosis information if the test fails (`TEST_ASSERT` branches to the `exit` label if the assertion fails).

mazimkhan commented:
This PR is critical as CI often fails in timing tests.

mazimkhan commented:
Running it in Oulu CI as that is most time skewed.

mazimkhan commented:
Timing test fails on Oulu CI https://jenkins-internal.mbed.com/job/mbedtls-pr/314/execution/node/206/log/

mpg commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,362 @@
+exit:
+    /* No cleanup needed, but show some diagnostic iterations, because timing
+       problems can be hard to reproduce. */
+    mbedtls_fprintf( stdout, "  Finished with millis=%lu get(timer)<=%lu iterations=%lu\n",
```
Thanks for including those diagnostic messages! However, I seem to remember that stdout from tests is filtered out by the framework unless `--verbose` is used (a decision I could never understand by the way). Is that still the case? If so, we should ensure that `--verbose` is always used when running the tests on the CI.

mpg commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,362 @@
+    /* No cleanup needed, but show some diagnostic iterations, because timing
+       problems can be hard to reproduce. */
+    mbedtls_fprintf( stdout, "  Finished with millis=%lu get(timer)<=%lu iterations=%lu\n",
+                     millis, mbedtls_timing_get_timer( &timer, 0 ),
```
I'm wondering if we should include `new_milis` too, in case the failing assertion is `new_millis >= millis`, and the return value of `get_timer()` happens to change just between the `TEST_ASSERT` and the `printf()`.

mpg commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,362 @@
+    unsigned long millis = 0;
+    unsigned long iterations = 0;
+
+    /* Start the timer. After setting it for the first time, the return value
```
Should it be _when_ setting it for the first time, rather that after?

mpg commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,362 @@
+        /* The timers shouldn't drift apart, i.e. millis2-millis1 should stay
+           roughly constant, but this is hard to test reliably, especially in
+           a busy environment such as an overloaded continuous integration
+           system, so we don't test it it. */
```
Typo: it it.

mpg commented on tests/suites/test_suite_timing.function:
```diff
@@ -1,15 +1,362 @@
+       the alarm (with second resolution) and the timer (with millisecond
+       resolution) are based on different clocks. Be very liberal with the
+       upper bound, because the platform might be busy. */
+     unsigned long millis_min = ( seconds > 0 ?
```
Style: indented with 5 spaces.

gilles-peskine-arm commented on library/timing.c:
```diff
@@ -378,13 +383,16 @@ static void busy_msleep( unsigned long msec )
+        if( verbose != 0 )                                              \
+        {                                                               \
+            mbedtls_printf( "failed at line %d\n", __LINE__ );          \
+            mbedtls_printf( " cycles=%lu ratio=%lu millisecs=%lu secs=%lu hardfail=%d a=%lu b=%lu\n", cycles, ratio, millisecs, secs, hardfail, (unsigned long) a, (unsigned long) b ); \
```
Done.

gilles-peskine-arm commented:
The CI has run several times, but due to a Jenkins issue it picked old revisions of this PR. Manual run on the latest version (https://github.com/ARMmbed/mbedtls/pull/1136/commits/ca5f6e2290735f63804ff58c4f01070c502db148): https://jenkins-internal.mbed.com/job/mbedtls-commit-tests/47/

gilles-peskine-arm commented:
I uploaded a fix for a mistake that Clang pointed out.

gilles-peskine-arm commented:
New bug fix: the code of `mbedtls_timing_get_timer` was originally using an uninitialized integer variable on Windows when calling it to initialize a timer, and I introduced a similar bug on Unix. I've now changed it to not read any part of the timer object when resetting.

This bug is harmless on basically all real-world environments since the worst it could do was to cause an integer overflow in an expression whose value was not specified. Nonetheless, it could cause a crash if someone compiled a Windows or Unix application with signed integer overflow checks enabled.

gilles-peskine-arm commented:
CI release job: https://jenkins-internal.mbed.com/job/mbedtls-release/51/

CI commit job: https://jenkins-internal.mbed.com/job/mbedtls-commit-tests/49/

gilles-peskine-arm commented:
CI commit job: https://jenkins-internal.mbed.com/job/mbedtls-commit-tests/52/

CI release job: https://jenkins-internal.mbed.com/job/mbedtls-release/60/

gilles-peskine-arm commented:
I've further relaxed the self-test. CI commit job: https://jenkins-internal.mbed.com/job/mbedtls-commit-tests/55/

gilles-peskine-arm commented:
PR job https://jenkins-internal.mbed.com/job/mbedtls-pr/525/ failed only the self-test on FreeBSD (1 out of 2). Commit job https://jenkins-internal.mbed.com/job/mbedtls-commit-tests/55/ passed the self-test on FreeBSD both times (and failed in the other timing issue, in `compat.sh`, but that's unrelated with this PR).

mpg commented:
I'd like to advocate two things: (1) that we merge this now rather than later. Even if it's not perfect, it's an improvement over the current state of things, so I'd rather have an improvement now, and a further improvement later, that maybe something perfect some distant time in the future. (2) I think we should skip the timing selftest function in the CI until further notice (eg until the CI machines get less loaded): as discussed, the timing functions are loosely coupled with the rest of the library and the implementations we provide are likely to be replaced with other (user-provided) implementations in production code, so the looser tests introduced by this PR should be enough.

gilles-peskine-arm commented:
@mpg Agreed about merging ASAP. I've been fixing some build failures under various compiler warning settings, but according to CI this has converged. Even if it isn't perfect, it's already a major improvement.

For dropping the timing self-test, what's your proposed approach? New option to `selftest`?

mpg commented:
For dropping the timing selftest, I was thinking we could just stop running `programs/tests/selftest` as all the other selftests are already included in their corresponding test suite.

mpg commented:
Also, regarding backports, I'd like to suggest skipping the backport to 1.3 as it's only going to see one more release before being end-of-lifed, so I don't see the point in backporting improvements to tests to that branch.

gilles-peskine-arm commented:
Following our discussion, I'm going to rebase this PR to clean up the history (no reversal in the changes to mbedtls_timing_get_timer, less waffling in the evolution of the tests).

gilles-peskine-arm commented:
Rebased with a cleaner history. The old history is at https://github.com/gilles-peskine-arm/mbedtls/tree/IOTSSL-1798-development-2

mpg commented:
I've reviewed the diff between the previous version of this branch (in my local copy) and the rebased version, and I'm satisfied nothing nasty happened during the rebase.

gilles-peskine-arm commented:
I triggered a daytime run of the CI job that tends to fail: https://jenkins-internal.mbed.com/job/mbedtls-commit-tests/56/ . All the timing tests passed, there were failures only in compat.sh which is a known issue (also due to CI slowness) that wasn't fixed when this PR was forked.

mazimkhan commented:
retest

mazimkhan commented:
retest
